File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        if(!arr[start])
            return true;
        if(arr[start] == -1)
            return false;
        
        int right = arr[start] + start, left = start - arr[start];
        arr[start] = -1;
        
        return ((right < arr.size() && canReach(arr, right)) || (left >= 0 && canReach(arr, left)));
    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.go
func canReach(arr []int, start int) bool {
    if arr[start] == 0 {
        return true
    }
    var q []int
    vis := make([]bool, len(arr))
    vis[start] = true
    q = enQ(arr, q, start)
    for pos:=0; pos < len(q); pos++ {
        curr := q[pos]
        if vis[curr] {
            continue
        }
        vis[curr] = true
        if arr[curr] == 0 {
            return true
        }
        q = enQ(arr, q, curr)
    }
    return false
}

func enQ(arr, q []int, i int) []int {
    if i - arr[i] >= 0 {
        q = append(q, i - arr[i])
    }
    if i + arr[i] < len(arr) {
        q = append(q, i + arr[i])
    }
    return q
}